src/Servers/Kestrel/Core/src/Internal/Http3/Http3Connection.cs

```diff
@@ -286,7 +270,19 @@ private void UpdateStartingStreams(DateTimeOffset now)
                             // Request stream
                             UpdateHighestStreamId(streamIdFeature.StreamId);
 
-                            var stream = new Http3Stream<TContext>(application, httpConnectionContext);
+                            // Check whether there is an existing HTTP/3 stream on the transport stream.
+                            // A stream will only be cached if the transport stream itself is reused.
+                            var stream = streamContext.Features.Get<ICachedHttp3StreamFeature<TContext>>()?.CachedStream;
+                            if (stream == null)
+                            {
+                                stream = new Http3Stream<TContext>(application, CreateHttpStreamContext(streamContext));
+                                streamContext.Features.Set<ICachedHttp3StreamFeature<TContext>>(new DefaultCachedHttp3StreamFeature<TContext>(stream));
```

Comment:
The feature collection doesn't make sense as a caching mechanism since the collection itself shouldn't preserve state between uses. You'd want the lower layer to provide an explicate caching feature.

Reply:
Do you have a suggestion about how that might work? Would it be a new feature with a dictionary that gets preserved between requests like:

```cs
public interface IStateFeature
{
   IDictionary<string, object> State { get; }
}
```

Or an explicit dictionary on transport context?

I created an issue a while back about using features to preserve the state between requests - https://github.com/dotnet/aspnetcore/issues/6895. There is a lot of crossover between that and this, just at a different layer.

I don't want to turn this into a big thing so another option is to have a hack in .NET 6 where the stream is added to the transport context's `Items` with a specific key and that specific key isn't cleared between calls.

Reply:
The state feature and the dictionary on the transport context accomplish the same end, I don't have a preference between them.

> I don't want to turn this into a big thing so another option is to have a hack in .NET 6 where the stream is added to the transport context's `Items` with a specific key and that specific key isn't cleared between calls by the QUIC transport. We could then improve it in .NET 7.

That is hackier than I want to go. Given our http/3 goals for 6.0 emphasize functionality over perf, I'd rather not cache if we can't come up with a clear mechanic for it.